### DISCUSSION TITLE
How do I use NuGet or other third party plugins in RavenDB?

### DISCUSSION THREAD

**Discussion by Pichardescu** [ravendb/ravendb#14949] [Q&A] [answered]
Hi!

I found this two NuGet packages with functions that I'm going to need for my next project in **NodeJS**:

- https://www.nuget.org/packages/RavenDB.Client.UniqueConstraints/3.5.9
- https://www.nuget.org/packages/RavenDB.Bundles.CascadeDelete/3.5.9

But there is no documentation anywhere:
- Do they work in the RavenDB Server? Or they're just for C# Applications?
- Can I make them work globally in the server, always for any kind of query? Or they are just for indexes?
- If they work only with indexes, it's enough to just add an "Additional Assembly" with the Package Name and Package Version?
    - Do I need to install those packages globally in my PC?
    - Do I need to add the files to a special folder?
    - Do I need to change the RavenDB settings in some way?
- How do I use them?
    - Do I need to write something 

Thanks!

## Background: No need to read

I'm looking for a Database for my next project in NodeJS.

I'm deciding between RavenDB and MongoDB, so far RavenDB looks gorgeous in comparison, except for the lack of third-party utilities that I consider essential for my project.

I'm aware that RavenDB doesn't need an ORM, but it's a great PLUS, i.e., in the case of MongoDB we have "Mongoose" and "PRISMA".

Apart from all the features and comparisons out there, these are the main features I like.

Mongoose and Prisma offers:

- Pretty Schema's and Types Validation that doesn't interfere with scalability.
- **Unique Constraints** <
- A beautiful way of querying with NodeJS, with snippets and schema-based hints:
`const allUsers = await prisma.user.findMany()`

Mongoose:

- mongoose-mpath: An awesome third-party plugin that allows me to work and handle infinite nested trees very easily with **Referential Integrity**.

Prisma:

- **Simulated Referential Integrity** <

With RavenDB I know that I'm saying goodbye to an effortless way of working in exchange for more control.

But at least I need the "Unique Constraints" and "Simulated Referential Integrity".

**Comment by ayende** [comment, marked as answer]
Both of those packages are for RavenDB 3.5, which is no longer supported.

At any rate, you can do most of what you wnat in different manners.
For enabling globally, see the events on the node.js client, which will let you handle things in a corss cutting manner.

The Additional Assembly allows you to execute interesting things during indexing time, I don't think that this is applicable for your scenario.

RavenDB *has* an OR/M, it is just that you get that by default from the client. 

Note that unique constraints in RavenDB would work differently (need cluster wide transactions for that, so require explciit behavior), see: https://ayende.com/blog/194405-A/ravendb-5-2-simplifying-atomic-cluster-wide-transactions

For queries, note that we provide support for this. So you can do:

```
var user = session.query<User>()
         .where("name", name)
         .single();
```

And you'll get proper autocomplete in the `where(|` portion

**Reply by Pichardescu** [reply]
Thank you very much! That answers all my questions.

With your response and after studying the bootcamp course in Github, now I'm in synthony with RavenDB, I had a bad start studying the documentation and YouTube videos first, but now I understand everything perfectly.

Again, thank you very much.
